### ISSUE TITLE
staff break moving when downloading

### ISSUE THREAD

**Issue by annamorphism** [DDMAL/Neon#762]
In some circumstances, downloading the MEI file seems to result in an incorrect MEI file, which then has many (hard-to-correct) mistakes after uploading. For example, after folio 18v of Einsiedeln had been corrected, downloaded, and re-uploaded, the result looked like this:

![image](https://user-images.githubusercontent.com/67451875/153500219-13137ba5-cb2b-4ba9-a085-7cfd0c28577d.png)

In the MEI , the sb marker that should be breaking this lengthy syllable has been moved to the end, instead of the two halves of the syllable being linked across it. As a result, the neumes loop back to the beginning of the staff, creating all sorts of chaos. 
![image](https://user-images.githubusercontent.com/67451875/153500418-75149121-8b04-4260-8ced-b4fb418ade61.png)

Since the chaos only appears after one has downloaded and re-uploaded the file, it follows that the bug is somewhere in the "download file" process.

**Comment by JoyfulGen** [comment]
Additional information: it looks like the problem arises when there are divisios within a syllable that is also linked with the syllable on the next staff (which explains why we never encountered this problem in Salzinnes). From what I understand, when there are divisios within one syllable, they all appear after the closing `<syllable>` in the MEI file, like so:
<img width="266" alt="Divisios in syl 018v" src="https://user-images.githubusercontent.com/83373378/153905730-e3be9c81-4d00-4ca9-b6c6-4ff47d705a6f.png">
<img width="1018" alt="Normal divisio within syllable behaviour" src="https://user-images.githubusercontent.com/83373378/153905745-767508b0-e7e6-4c32-ba1c-b1167aeecede.png">

However, when a syllable containing divisios is toggled with a syllable on the following staff, the program wants to put the divisios after the final `<syllable>`, so it puts the `<sb>` after that, which makes it think that the toggled syllables are actually one same syllable on the same staff. That's what results in the dramatic situation seen in Anna's screenshot above. Folio 018v has another example of this that's a bit smaller, so easier to get in one screenshot:
<img width="767" alt="Divisio in toggled syl 018v" src="https://user-images.githubusercontent.com/83373378/153906839-a39b3598-9816-4124-b5fc-35f7f8203cac.png">
<img width="972" alt="Divisio in toggled syllable chaos" src="https://user-images.githubusercontent.com/83373378/153907418-1e7794c8-0b70-4f61-8d89-b1e1d28d372b.png">

I  don't believe there's a fix for this at the moment. As many times as we put the neumes back, if we try to toggle the syllables the resulting MEI file will interpret them as one single syllable. So we currently have no way of inputting a syllable that contains divisios and is toggled with another syllable over the staff break.

Screenshots of MEI come from this file: [CH-E_611_018v.mei.zip](https://github.com/DDMAL/Neon/files/8062503/CH-E_611_018v.mei.zip)

**Comment by yinanazhou** [comment]
This issue is caused by the insert action of divisio. When one insert a divisio, the divisio is added as a child of the `layer` element. I need some support from @annamorphism and @JoyfulGen, where would a divisio appear? In most cases, it's a child of `layer`. In this case, it's a child of a `syllable`. Is there any other possibilities?

**Comment by yinanazhou** [comment]
In addition, the same issue may also happen to `accid`. What are the possible positions of `accid`?

**Comment by annamorphism** [comment]
@yinanazhou I think the divisio can only be a child of layer and syllable; it wouldn't make sense to divide a neume with a div-line, for example. On the other hand, you might be able to have multiple accidentals in the same syllable applying to different neumes (I don't think it happens much, but it could in theory.) Does that help?

**Comment by yinanazhou** [comment]
I have already merged the changes. Now you should be able to insert an accid or divisio inside a syllable. Please let me know if it fixes the issue! Thanks!

**Comment by annamorphism** [comment]
@yinanazhou you can now put div-lines inside a syllable BUT putting accid inside a syllable leads to an invalid MEI flag.

**Comment by yinanazhou** [comment]
Hi @annamorphism, this is in the MEI schema. To make accid a valid child of syllables, I will need to modify the MEI schema. I will have to make sure that accids do go into the syllable. Could you please provide an example folio where accid is in the syllable?

**Comment by annamorphism** [comment]
@yinanazhou I was working with the (invalid) generated file for 26r of Einsiedeln, which has an accidental as the child of the syllable. BUT in that case you don't *need* it to be--there's nothing about the situation preventing the accidental from being up a level. I can look around for a case where it needs to be a child of syllable.

**Comment by annamorphism** [comment]
@yinanazhou I recommend folio 17v, which has a flat in the middle of a syllable:
![image](https://user-images.githubusercontent.com/67451875/155025796-94b9cdf6-757e-495a-91df-a376b389f262.png)

and also 12r, where there is a flat as part of a syllable linked across lines. 
![image](https://user-images.githubusercontent.com/67451875/155025935-f2884839-73c5-4f2e-a807-f2dbc0ede27f.png)

 Right now 17v is actually valid, but if I take the flat out and put it back in in the same place, it is invalid. 
And 12r is technically valid, but has the same problems of things being on the wrong staff that was being a problem with div-line. 
Both of these are taken from the "processed" file in omr-resources.

**Comment by yinanazhou** [comment]
@annamorphism I checked 017v. The MEI file was valid at first because the `accid` element is not inside the `syllable`. But when you insert it again in Neon, Neon automatically detects that it should be inside the syllable, so the accid is inserted inside the syllable, thus making the MEI file invalid. 

I will use this folio as an example to make a pull request to modify the class of `accid` in mei-encoding. @carrieeex could you work on this during the reading week?

**Comment by JoyfulGen** [comment]
Inserting a divisio in a syllable appears to work! If I input a divisio between two neumes that are part of the same syllable, it becomes part of that syllable. We can now toggle syllables that contain divisios!

I think, however, that this only works if the divisios are input into the syllable after it's been grouped. If I do the opposite, by merging syllables around divisios that are already on the page, the divisios don't become part of the new syllable. I have to delete them, and reinput them. @yinanazhou, would it be possible to change that, so that divisios that are already on the page automatically become part of a syllable that's merged around them?

**Comment by JoyfulGen** [comment]
I've found an issue with the divisio in syllables. If a divisio is inserted between two neumes that are part of the same syllable, the fix works perfectly. However, there are situations (such as in the screenshot below) where the syllable at the end of a staff is followed by a divisio, but no other neume after that. In that situation, since there's only a neume before the divisio and not after, the divisio doesn't become part of the syllable preceding it and the syllable can't be toggled with the syllable on the next staff.

Is there a way that a divisio could also become part of a syllable when it's surrounded by two neumes that are part of a toggled syllable?

<img width="934" alt="006r divisio after neume example" src="https://user-images.githubusercontent.com/83373378/155154878-676ce1fe-99c3-448a-a31c-a06d2ac256e1.png">

**Comment by JoyfulGen** [comment]
I've found another little problem with divisios that are children of `syllable`. If I ungroup a syllable that contains divisios, each of those individual divisios becomes its own syllable, with its own bounding box and text box. This should not happen, because a single divisio can't be sung or contain text. A demonstration where, after I ungroup the giant syllable, when I click on a divisio, its bounding box becomes highlighted (please disregard the overwhelmingly bad result of the OMR here, the script is completely different and the OMR can't handle it):

https://user-images.githubusercontent.com/83373378/155560533-55dbef32-30ee-48f3-88f2-660fbe6ba55a.mov

This is what the divisio/syllable element looks like in the MEI:

<img width="884" alt="281v_divisio_syllable" src="https://user-images.githubusercontent.com/83373378/155560907-33a53ee6-5d01-42a1-b581-4e77090c92b9.png">

It looks like a perfectly normal syllable to me, except that instead of a neume there's a `divLine`.

The complete MEI file: 
[281v_divisio_is_syllable.mei.mei.zip](https://github.com/DDMAL/Neon/files/8133955/281v_divisio_is_syllable.mei.mei.zip)
(beep, boop, woosh, woo, and donk are the syllables that are divisios, if you want to find them quickly)

Could a possible fix be to make it so when a syllable containing divisios is ungrouped, those divisios become children of `layer` again?

**Comment by yinanazhou** [comment]
> Inserting a divisio in a syllable appears to work! If I input a divisio between two neumes that are part of the same syllable, it becomes part of that syllable. We can now toggle syllables that contain divisios!
> 
> I think, however, that this only works if the divisios are input into the syllable after it's been grouped. If I do the opposite, by merging syllables around divisios that are already on the page, the divisios don't become part of the new syllable. I have to delete them, and reinput them. @yinanazhou, would it be possible to change that, so that divisios that are already on the page automatically become part of a syllable that's merged around them?

Hi @JoyfulGen , to make divLine automatically added into a syllable, I will need to modify the mei-encoding job in Rodan. For now, I propose a new option called "Insert to nearest syllable" for accidentals and divLines. The general workflow is, when you open a new mei file in neon, all the divLines and accidentals are not in the syllables. You will need to click-select them, and click the `Insert to nearest syllable button`. Neon will insert the selected element into the nearest syllable at the right position. Do notice that this option is only one-way for now. If you need an option that can revert the insertion other than the UNDO button, please let me know. 

In addition, with this new option, you should be able to merge syllables with divLines or accidentals in between. Then, all you need to do is to select the divLine or accidental and click the `Insert to nearest syllable` button. This workflow can go around the group-with-divLine-and-accidental problem.

I've tested it and it worked well. Please let me know if anything unexpected happens! Thanks!

**Comment by JoyfulGen** [comment]
It works perfectly so far! I'm now correcting the folios I'd already worked on that had a divisio within a toggled syllable (for reference: 006r, 007v, 008r, 010r (2), 011v, 012r (3), 012v (2), 016r (2), 018r, 020r). I'll update you when I'm done!

**Comment by JoyfulGen** [comment]
UPDATE for @annamorphism and @yinanazhou 

So I think this issue has actually been fixed by https://github.com/DDMAL/Neon/issues/676?!? 

I've been messing around with the above-mentioned folios and I realized that toggling a syllable with a divisio inside it now always works, regardless of whether that divisio is a child of `layer` or `syllable`. I believe this is because it used to be that when a syllable was toggled, the two parts of the syllable behaved as if they were merged: there were no `@precedes` or `@follows` and the `<clef>`, `<custos>` and `<sb>` elements were part of that syllable. When there was a divisio, it went to the very end of that weird syllable and that's what caused the chaos. But in https://github.com/DDMAL/Neon/issues/676, @yinanazhou fixed it so that toggled syllables remain separate, but have the `@precedes` or `@follows` added to them. This fixes the divisio chaos issue entirely, because even if the divisio is a child of `layer` and goes after the closing `<syllable>` element, the two syllables toggled are still separate, so the divisio goes after the ***first*** one and doesn't interfere with the toggling at all! Here's a demo:

In toggled syllable "heya," the `divLine` is a child of `syllable`, which is the fix proposed here. It works very well, the end result in Neon is a perfect and lovely toggled syllable:

<img width="931" alt="010r child of syllable" src="https://user-images.githubusercontent.com/83373378/156593297-9f3dcbad-375e-4984-9477-9841c3a81b75.png">

In toggled syllable "hello," the `divLine` is a child of `layer`, and it also works perfectly, because it appears after the first syllable, not after the whole thing:

<img width="953" alt="010r child of layer" src="https://user-images.githubusercontent.com/83373378/156593469-e69dab3a-1eac-424a-9166-dc208158a24b.png">

The point of all this being that even though this fix appears to work, it's possibly become superfluous. @annamorphism, if the toggling isn't a problem anymore, is there still a reason for a divisio or an accidental to be part of a syllable?

The complete MEI file: [CH-E_611_010r_heyahello.mei.zip](https://github.com/DDMAL/Neon/files/8178908/CH-E_611_010r_heyahello.mei.zip)

**Comment by annamorphism** [comment]
@JoyfulGen you might be right--I think we might  not need the divisios to be part of a syllable, if the toggling works. I'm not as sure about accidentals, because they have (or can have) an effect on the syllable in a way divisios don't.

**Comment by yinanazhou** [comment]
Hi @annamorphism and @JoyfulGen , I've pushed the changes to automatically update `@precedes` and `@follows` after grouping neumes. Both this issue and issue #676 should be fixed. Please let me know if anything unexpected happens.

I've deleted the function that automatically inserts divLine/accidental inside syllable. I didn't delete the `insert to nearest syllable` feature because MEI encoding is semantic. Although the position of divLine can be indicated by its facsimile zone so it looks good in Neon, it makes more sense to insert divLine into the syllable because the function of divLine is to separate neumes in the syllable. I'm not sure about the accidentals though. Please let me know if we still need accidentals inside syllables.

**Comment by JoyfulGen** [comment]
> It works perfectly so far! I'm now correcting the folios I'd already worked on that had a divisio within a toggled syllable (for reference: 006r, 007v, 008r, 010r (2), 011v, 012r (3), 012v (2), 016r (2), 018r, 020r). I'll update you when I'm done!

UPDATE: I was able to fix all of these without any issues. There are no syllables containing only divisios, divisios/accidentals don't make files invalid, and the `insert to nearest syllable` feature works!

**Comment by yinanazhou** [comment]
Hi @JoyfulGen and @annamorphism , could you please confirm that this issue can be closed?

**Comment by JoyfulGen** [comment]
I've found two small issues remaining, both related to the fact that divisios and accidentals are still being automatically inserted into syllables. The automatic insertion doesn't happen when I interact with the folio, but it's still happening when I upload a new file. It causes these issues:
-  When an accidental is part of a syllable, the file becomes invalid. 
- Sometimes the automatic function groups a syllable with the divisio before it. In these cases, so if the divisio is a child of a syllable and at the beginning of it, ungrouping the syllable makes the divisio into its own syllable, with a `syl` and a bounding box.

Would it be possible to also delete the function that automatically inserts divisios and accidentals into syllables when a file is uploaded to Neon? I think that would fix the remaining issues.

**Comment by yinanazhou** [comment]
Hi @JoyfulGen, I'm having trouble reproducing the automatic insertion of divLine into syllables. Could you please provide more details about it?

**Comment by yinanazhou** [comment]
Hi @JoyfulGen, I have pushed the changes for ungrouping when the syllable starts with a divLine or accidental. Ungrouping should be good now. I still cannot reproduce the automatic insertion when uploading issue though.

**Comment by JoyfulGen** [comment]
> Hi @JoyfulGen, I'm having trouble reproducing the automatic insertion of divLine into syllables. Could you please provide more details about it?

Absolutely! The automatic insertion never happens if I'm the one adding divLines or accidentals. The issue I'm noticing is that when I upload a new file to Neon, some divLines are already inserted into syllables (I'm assuming this is due to an automatic function, but I may be misunderstanding, sorry!). Here's an example from folio 266v, which I captured without having interacted with the folio at all: you can tell the divisios are part of the syllable surrounding them, because when I click on them the neumes become highlighted.

https://user-images.githubusercontent.com/83373378/160608959-02aef1e7-bcf3-4179-a14f-88872daf4176.mov

Here's the MEI file: 
[266vdivtest.mei.zip](https://github.com/DDMAL/Neon/files/8371265/266vdivtest.mei.zip)
I renamed the syllables containing divisios "zero" through "thirteen" to make them easier to find.

**Comment by yinanazhou** [comment]
Hi @JoyfulGen, I've failed to reproduce this issue both locally and in browser Neon. I first moved out all the divLines, and uploaded to Neon. But I could not reproduce the automatic insertion. Is it possible that the divLines are already inside syllables before uploading? Is the MEI file uploaded to Neon are the output of rodan jobs? If so, this might be in the MEI-encoding job. 

Although I cannot reproduce this issue and don't know its causes, I have two possible temporary solutions. 
1. add a check in Neon that automatically moves out divLines and accidentals when uploading MEI files.
2. add a button that can move out the selected divLine or accidental.

Please let me know which one you would prefer.

**Comment by JoyfulGen** [comment]
You're right, the divLines are already in the syllables before uploading! I'm sorry I missed that. I think solution 1 is ideal, because divLines that are actually supposed to be in syllables are quite rare, so adding them ourselves is no problem.
